<launch>

    <!-- Configurable arguments -->
    <arg name="robot_name" default="sam"/>

    <arg name="wp_tolerance" default="5."/>
    <arg name="depth_tolerance" default="1."/>
    <arg name="base_frame" default="$(arg robot_name)/base_link"/>
    <arg name="forward_rpm" default="400"/>
	

    <!-- topics, note the root! -->
    <arg name="rpm1_cmd_topic" default="/$(arg robot_name)/core/thruster1_cmd" />
    <arg name="rpm2_cmd_topic" default="/$(arg robot_name)/core/thruster2_cmd" />
    <arg name="as_rpm1_topic" default="/$(arg robot_name)/ctrl/goto_waypoint/rpm1" />
    <arg name="as_rpm2_topic" default="/$(arg robot_name)/ctrl/goto_waypoint/rpm2" />
    <arg name="rpm_enable_topic" default="/$(arg robot_name)/ctrl/goto_waypoint/rpm/enable" />

    <arg name="thrust_vector_cmd_topic" default="/$(arg robot_name)/core/thrust_vector_cmd" />
    <arg name="heading_setpoint_topic" default="/$(arg robot_name)/ctrl/yaw_setpoint" />
    <arg name="depth_setpoint_topic" default="/$(arg robot_name)/ctrl/depth_setpoint" />
    <arg name="vbs_setpoint_topic" default="/$(arg robot_name)/ctrl/depth_setpoint" />
    <arg name="vel_setpoint_topic" default="/$(arg robot_name)/ctrl/speed_setpoint" />
    <arg name="roll_setpoint_topic" default="/$(arg robot_name)/ctrl/roll_setpoint" />
    <arg name="yaw_feedback_topic" default="/$(arg robot_name)/dr/yaw" />
	<arg name="vel_feedback_topic" default="/$(arg robot_name)/dr/u" />
	<!--arg name="yaw_feedback_topic" default="/$(arg robot_name)/ctrl/odom_listener/yaw_feedback" /-->

	<!-- Services -->
	<arg name="toggle_yaw_ctrl_service" default="/$(arg robot_name)/ctrl/toggle_yaw_ctrl" />
	<arg name="toggle_depth_ctrl_service" default="/$(arg robot_name)/ctrl/toggle_depth_ctrl" />
    <arg name="toggle_vbs_ctrl_service" default="/$(arg robot_name)/ctrl/toggle_vbs_ctrl" />
	<arg name="toggle_speed_ctrl_service" default="/$(arg robot_name)/ctrl/toggle_speed_ctrl" />
	<arg name="toggle_roll_ctrl_service" default="/$(arg robot_name)/ctrl/toggle_roll_ctrl" />
    

    

    <node name="goto_waypoint" pkg="sam_action_servers" type="wp_depth_action_planner.py" output="screen" ns="$(arg robot_name)/ctrl">
		<param name="wp_tolerance" value="$(arg wp_tolerance)"/>
		<param name="depth_tolerance" value="$(arg depth_tolerance)"/>
		<param name="base_frame" value="$(arg base_frame)"/>
		<param name="base_frame_2d" value="$(arg robot_name)/base_link_2d"/>
		<param name="forward_rpm" value="$(arg forward_rpm)"/>
		
		<param name="rpm1_cmd_topic" value="$(arg as_rpm1_topic)" />
        	<param name="rpm2_cmd_topic" value="$(arg as_rpm2_topic)" />
		<param name="heading_setpoint_topic" value="$(arg heading_setpoint_topic)" />
		<param name="depth_setpoint_topic" value="$(arg depth_setpoint_topic)" />
		
		<!--Turbo turn parameters-->
  		<param name="turbo_turn_flag" value="False" />
		<param name="thrust_vector_cmd_topic" value="$(arg thrust_vector_cmd_topic)" />
		<param name="yaw_feedback_topic" value="$(arg yaw_feedback_topic)" />
		<param name="turbo_angle_min_deg" value="90.0" />		
		<param name="flip_rate" value="0.5" />
		<param name="rudder_angle" value="0.1" />
		<param name="turbo_turn_rpm" value="1000" />
		<param name="vbs_setpoint_topic" value="$(arg vbs_setpoint_topic)" />
		
		<!--Vel ctrl parameters-->
  		<!--param name="vel_ctrl_flag" value="False" />
		<param name="vel_setpoint" value="0.5" /-->
		<param name="roll_setpoint" value="0.0" />
		<param name="vel_setpoint_topic" value="$(arg vel_setpoint_topic)" />
		<param name="roll_setpoint_topic" value="$(arg roll_setpoint_topic)" />
		<param name="vel_feedback_topic" value="$(arg vel_feedback_topic)" />
		<!--Hard override vel ctrl -->
		<param name="use_constant_rpm" value="False" /> 
		<param name="rpm_enable_topic" value="$(arg rpm_enable_topic)" />

		<!--WP following augmentations for using VBS, crosstrack error and WP overshoot handling-->
  		<param name="vbs_diving_flag" value="True" />
  		<param name="crosstrack_flag" value="True" />
		<param name="lookahead_dist" value="3.0" />
  		<param name="wp_overshoot_flag" value="True" />
  		<param name="timeout_flag" value="True" />
		<param name="timeout_limit" value="500" />


		<!--Controller services-->
		<param name="toggle_yaw_ctrl_service" value="$(arg toggle_yaw_ctrl_service)" />
		<param name="toggle_depth_ctrl_service" value="$(arg toggle_depth_ctrl_service)" />
		<param name="toggle_vbs_ctrl_service" value="$(arg toggle_vbs_ctrl_service)" />
		<param name="toggle_speed_ctrl_service" value="$(arg toggle_speed_ctrl_service)" />
		<param name="toggle_roll_ctrl_service" value="$(arg toggle_roll_ctrl_service)" />


	</node>

    <node name="rpm_repub" pkg="sam_action_servers" type="rpm_repub.py" output="screen" ns="$(arg robot_name)/ctrl">
		<param name="as_rpm1_topic" value="$(arg as_rpm1_topic)" />
        	<param name="as_rpm2_topic" value="$(arg as_rpm2_topic)" />
		<param name="rpm1_cmd_topic" value="$(arg rpm1_cmd_topic)" />
        	<param name="rpm2_cmd_topic" value="$(arg rpm2_cmd_topic)" />
		<param name="rpm_enable_topic" value="$(arg rpm_enable_topic)" />
    </node>
</launch>
